src/Microsoft.AspNetCore.Hosting/WebHostExtensions.cs

```diff
@@ -95,7 +153,17 @@ private static void Run(this IWebHost host, CancellationToken token, string shut
                 },
                 applicationLifetime);
 
-                applicationLifetime.ApplicationStopping.WaitHandle.WaitOne();
+                var waitForStop = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+                applicationLifetime.ApplicationStopping.Register(obj =>
+                {
+                    var tcs = (TaskCompletionSource<object>)obj;
+                    tcs.TrySetResult(null);
+                }, waitForStop);
+
+                await waitForStop.Task;
```

Comment:
Does this men you can call StopApplication from a request thread now? Also see https://github.com/aspnet/KestrelHttpServer/issues/1501 for an example where somebody is trying to implement graceful shutdown. Does this make it any easier?